### PR TITLE
DOCSP-16466: TLS warning and OCSP stapling

### PR DESCRIPTION
**Pull request by ccho-mongodb**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-16466

### Docs staging link (requires sign-in on MongoDB Corp SSO):

Note at the bottom of this section: https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-16466-ocsp-stapling/quick-start/#query-your-mongodb-cluster-from-your-application

https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-16466-ocsp-stapling/fundamentals/connection/tls/#client-driven-ocsp

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### If your page documents a concept, does it meet the following criteria?

- [x] Target the [Jasmin persona](https://drive.google.com/file/d/14FbBOLCVxwSP6M9BK4Nz1Ir9tzxT8_02/view)
- [x] Target the [Lucas persona](https://drive.google.com/file/d/1J2vqJxo7ldv7OP_obA9Q-avf0o_ju4Lk/view)

### CODE REVIEW

File: source/fundamentals/connection/tls.txt [hunk 1]
```diff
@@ -211,3 +211,82 @@ object to the builder in the ``applyToSslSettings()`` lambda:
...
+   :widths: 35 65
+
+   * - Property
+     - Value
+
+   * - ``com.sun.net.ssl.checkRevocation``
+     - Set this property to ``true`` to enable revocation checking.
+
+   * - ``ocsp.enable``
+     - Set this property to ``true`` to enable client-driven OCSP.
```
Comment (terakilobyte) [inline source/fundamentals/connection/tls.txt:256]: Question: Is there any default here, and if there is, would it be worth mentioning? This question applies to the other table in OCSP Stapling as well.

File: source/fundamentals/connection/tls.txt [hunk 2]
```diff
@@ -211,3 +211,82 @@ object to the builder in the ``applyToSslSettings()`` lambda:
...
+
+   If the OCSP responder is unavailable, the TLS support provided by the
+   JDK reports a "hard fail". This differs from the "soft fail" behavior of
+   the MongoDB Shell and certain drivers.
+
+OCSP Stapling
+~~~~~~~~~~~~~
+
+OCSP Stapling is a type of OCSP in which the server must obtain the signed
+certificate from the certificate authority (CA) and including it in a
```
Comment (terakilobyte) [inline source/fundamentals/connection/tls.txt:268]:
```suggestion
certificate from the certificate authority (CA) and include it in a
```

File: source/fundamentals/connection/tls.txt [hunk 3]
```diff
@@ -211,3 +211,82 @@ object to the builder in the ``applyToSslSettings()`` lambda:
...
+.. warning::
+
+   If the OCSP responder is unavailable, the TLS support provided by the
+   JDK reports a "hard fail". This differs from the "soft fail" behavior of
+   the MongoDB Shell and certain drivers.
+
+OCSP Stapling
+~~~~~~~~~~~~~
+
+OCSP Stapling is a type of OCSP in which the server must obtain the signed
```
Comment (terakilobyte) [inline source/fundamentals/connection/tls.txt:267]:
```suggestion
OCSP stapling is a type of OCSP in which the server must obtain the signed
```
suggestion for consistency. The [RFC](https://datatracker.ietf.org/doc/html/rfc6961#section-1) and the  wiki article on it uses lowercase 's'

File: source/includes/quick-start/force-tls.rst [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+      jdk.tls.client.protocols=TLSv1.2
+
+   If you run your application from the command line or your IDE, you can
+   specify this property as a JVM argument using the following syntax:
+
+   .. code-block:: properties
+
+      -Djdk.tls.client.protocols=TLSv1.2
+
```
Comment (terakilobyte) [inline source/includes/quick-start/force-tls.rst:27]: This is great. Should it also be added to the FAQ?

File: source/fundamentals/connection/tls.txt [hunk 5]
```diff
@@ -211,3 +211,82 @@ object to the builder in the ``applyToSslSettings()`` lambda:
                 })
         .build();
    MongoClient client = MongoClients.create(settings);
+
+Online Certificate Status Protocol (OCSP)
+-----------------------------------------
+
+OCSP is a standard used to check whether X.509 certificates are valid or
+revoked during the TLS handshake. The driver supports the following variations
+of OCSP:
```
Comment (terakilobyte) [inline source/fundamentals/connection/tls.txt:220]: suggestion: I wonder if the removal of the conjunction could simplify this?

```suggestion
OCSP is a standard used to check the validity of X.509 certificates during the TLS handshake. The driver supports the following variations
of OCSP:
```

**Review by terakilobyte — APPROVED** [review]
LGTM. A couple of nonblocking suggestions and questions.

File: source/fundamentals/connection/tls.txt (hunk 1 above)
Comment (ccho-mongodb) [reply source/fundamentals/connection/tls.txt:256]: No, there's no default value. The variable names are from the built-in J2SE handling of OCSP:
https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ocsp.html

File: source/fundamentals/connection/tls.txt (hunk 3 above)
Comment (ccho-mongodb) [reply source/fundamentals/connection/tls.txt:267]: Good catch!

File: source/includes/quick-start/force-tls.rst (hunk 4 above)
Comment (ccho-mongodb) [reply source/includes/quick-start/force-tls.rst:27]: Created a ticket: https://jira.mongodb.org/browse/DOCSP-17059

File: source/fundamentals/connection/tls.txt (hunk 5 above)
Comment (ccho-mongodb) [reply source/fundamentals/connection/tls.txt:220]: I think my usage of valid might not be specific enough. I was hoping that the juxtaposition of "valid" and "revoked" would signal opposites. OCSP as far as I know is only to check revocation status of a certificate with information that a revocation server can access. I don't know if revocation servers have any other responsibilities other than checking whether a certificate has been revoked prior to its expiry (whether the certificate is on the Certificate Revocation List), i.e. it doesn't explicitly validate anything about the certificate.
I found some information about the potential statuses that the revocation server can return, so I'll update.

**Comment by ccho-mongodb** [comment]
Hi @jyemin and @rozza , in this PR, I added info on OCSP and OCSP stapling. Could one of you take a look when you get the chance?

**Review by jyemin — CHANGES_REQUESTED** [review]
I just ran into this myself while trying to connection to a Serverless cluster with 11.0.5.  I think the bug is reported at https://bugs.openjdk.java.net/browse/JDK-8236039 and fixed in 11.0.7 patch release.  If that's the case can we instead recommend an upgrade to 11.0.7 or later?

**Comment by ccho-mongodb** [comment]
> I just ran into this myself while trying to connection to a Serverless cluster with 11.0.5. I think the bug is reported at https://bugs.openjdk.java.net/browse/JDK-8236039 and fixed in 11.0.7 patch release. If that's the case can we instead recommend an upgrade to 11.0.7 or later?

Sure, that works. From the openjdk ticket, it looks like this was also an issue with some releases of 13, 14, and 15 so I will need to add those as well.

File: source/fundamentals/connection/tls.txt [hunk 6]
```diff
@@ -211,3 +211,87 @@ object to the builder in the ``applyToSslSettings()`` lambda:
...
+     - Set this property to ``true`` to enable revocation checking.
+
+   * - ``ocsp.enable``
+     - Set this property to ``true`` to enable client-driven OCSP.
+
+.. warning::
+
+   If the OCSP responder is unavailable, the TLS support provided by the
+   JDK reports a "hard fail". This differs from the "soft fail" behavior of
+   the MongoDB Shell and certain drivers.
```
Comment (jyemin) [inline source/fundamentals/connection/tls.txt:267]: certain => some other

File: source/fundamentals/connection/tls.txt [hunk 7]
```diff
@@ -211,3 +211,87 @@ object to the builder in the ``applyToSslSettings()`` lambda:
...
+
+OCSP Stapling
+~~~~~~~~~~~~~
+
+OCSP stapling is a type of OCSP in which the server must obtain the signed
```
Comment (jyemin) [inline source/fundamentals/connection/tls.txt:272]: Not sure about `type of OCSP` phrasing.  Is it really a type, or just a mechanism?

**Review by jyemin — APPROVED** [review]
A couple of phrasing suggestions, which you can feel free to ignore.

LGTM

File: source/fundamentals/connection/tls.txt (hunk 7 above)
Comment (ccho-mongodb) [reply source/fundamentals/connection/tls.txt:272]: I meant "type" as in "variation". I guess it's probably ok to assume the reader can draw the connection with OCSP and describe it instead as "a mechanism in which..."

**Pull request merged**